File: tree/bst/delete.go
package bst

import "trouble_solver/tree"

func deleteNode(root *tree.TreeNode, key int) *tree.TreeNode {
	if root == nil {
		return nil
	}

	if root.Val == key {
		if root.Left == nil {
			return root.Right
		} else if root.Right == nil {
			return root.Left
		} else {
			// 两边都不为空，选择右子树的最左侧
			node := root.Right
			// 上面判断过了，cur不可能为nil
			for node.Left != nil {
				node = node.Left
			}
			// 循环过后，cur就是替换的节点
			node.Right = deleteNode(root.Right, node.Val) // 这里一定要删除，再进行后面的操作
			node.Left = root.Left
			return node
		}
	}

	if root.Val < key {
		root.Left = deleteNode(root.Left, key)
	} else {
		root.Right = deleteNode(root.Right, key)
	}
	return root
}

File: tree/bst/delete.java
package tree.bst;

import tree.TreeNode;

/**
 * leetcode 450 middle 删除二叉搜索树中的节点
 */
public class Delete {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) {
            return null;
        }

        // 使用右子树的最左节点来代替root
        if (root.val == key) {
            if (root.right == null) {
                return root.left;
            } else if (root.left == null) {
                return root.right;
            } else {
                TreeNode cur = root.right;
                while (cur.left != null) {
                    cur = cur.left;
                }

                cur.right = deleteNode(root.right, cur.val);
                cur.left = root.left;
                return cur;
            }
        }

        if (root.val < key) {
            root.right = deleteNode(root.right, key);
        } else if (root.val > key) {
            root.left = deleteNode(root.left, key);
        }

        return root;
    }
}
